sdk/python/feast/pyspark/launchers/standalone/local.py

```diff
@@ -22,9 +23,71 @@
     StreamIngestionJobParameters,
 )
 
-# In-memory cache of Spark jobs
-# This is necessary since we can't query Spark jobs in local mode
-JOB_CACHE: Dict[str, SparkJob] = {}
+
+class JobCache:
+    """In-memory cache of Spark jobs.
+
+    This is necessary since we can't query Spark jobs in local mode
+
+    """
+
+    # Map of job_id -> spark job
+    job_by_id: Dict[str, SparkJob]
+
+    # Map of job_id -> job_hash. The value can be None, indicating this job was
+    # manually created and Job Service isn't maintaining the state of this job
+    hash_by_id: Dict[str, Optional[str]]
+
+    # This reentrant lock is necessary for multi-threading access
+    lock: threading.RLock
+
+    def __init__(self):
+        self.job_by_id = {}
+        self.hash_by_id = {}
+        self.lock = threading.RLock()
+
+    def add_job(self, job_id: str, job_hash: Optional[str], job: SparkJob) -> None:
```

Comment:
@oavdeev Is't it possible for us to simplify things to `def add_job(self, job: SparkJob)`? And have the `id` and `hash` be a property of the job? It seems like we are doing a lot of bookkeeping here with state that is somewhere derived from each other (id and hash from job)

Reply:
It's possible - I'm working on it. I'm combining this idea with what I've discussed with Oleg before. Basically, job hash should be part of `StreamIngestionJob`. In this case a lot of things become simpler to do - e.g. launcher won't have the ugly new `list_jobs_by_hash` method anymore.